src/lib/components/FileUpload/FileUpload.tsx

```diff
@@ -1,23 +1,167 @@
-      <button className="file-upload__button" type="button">
-        Drag and drop files here or click to upload
-      </button>
+    <div
+      className={classNames("p-form__group p-form-validation", {
+        "is-error": !!error,
+      })}
+    >
+      {label && <Label id={labelId}>{label}</Label>}
+      {help && <p className="p-form-help-text">{help}</p>}
+      <div className="p-form__control">
+        {!maxFiles || files.length < maxFiles ? (
+          <div className="file-upload__wrapper">
+            <div
+              {...getRootProps()}
+              aria-labelledby={label ? labelId : undefined}
+              className="file-upload"
+              data-testid="file-upload"
+            >
+              <button className="file-upload__button" type="button">
+                Drag and drop files here or click to upload
+              </button>
+            </div>
+          </div>
+        ) : null}
+        {error ? (
+          <p className="p-form-validation__message">
+            <strong>Error: </strong>
+            {error}
+          </p>
+        ) : null}
+        <div className="file-upload__files-list">
+          {rejectedFiles &&
+            rejectedFiles.map((rejection) => (
+              <span className="is-error" key={rejection.file.name}>
+                <div className="file-upload__file is-rejected">
+                  {rejection.file.name}
+                  <Button
+                    appearance="base"
+                    className="file-upload__file-remove-button"
+                    onClick={() => removeRejectedFile(rejection)}
+                    type="button"
+                  >
+                    <Icon name="close">Remove file</Icon>
+                  </Button>
+                </div>
+                {rejection.errors.map((error) => (
+                  <p
+                    className="p-form-validation__message"
+                    key={`${rejection.file.name}-${error.code}`}
+                  >
+                    {error.message}
+                  </p>
+                ))}
+              </span>
+            ))}
+          {files &&
+            files.map((file) => (
+              <div className="file-upload__file" key={file.name}>
+                {file.name}
+                <Button
+                  appearance="base"
+                  className="file-upload__file-remove-button"
+                  onClick={() => removeFile(file)}
+                  type="button"
+                >
+                  <Icon name="close">Remove file</Icon>
+                </Button>
+              </div>
+            ))}
+        </div>
+      </div>
     </div>
   );
 };
+
+export const FileUploadContainer = ({
+  help,
+  label,
+  maxFiles,
+  maxSize,
+}: Pick<FileUploadProps, "help" | "label" | "maxFiles" | "maxSize">) => {
+  const [files, setFiles] = useState<File[]>([]);
+  const [rejectedFiles, setRejectedFiles] = useState<FileRejection[]>([]);
+  const [error, setError] = useState("");
```

Comment:
Managing state of this component involves multiple pieces values where the next state depends on the previous one and it's not represented clearly or efficiently by using multiple useState hooks. I'd like you to challenge you and suggest you try a different approach. `useReducer` would simplify state updates and clearly describe how the state changes based on each action.

Below is just a rough example.

```tsx
const fileUploadReducer = (state, action) => {
  switch (action.type) {
    case "ADD_FILES": {
      const { acceptedFiles, fileRejections } = action.payload;
      return {
        ...state,
        files: [...state.files, ...acceptedFiles],
        rejectedFiles: [...state.rejectedFiles, ...fileRejections],
        error: fileRejections.length > 0 ? "Some files were rejected." : "",
      };
    } [...]
```

```suggestion
  const [state, dispatch] = useReducer(fileUploadReducer, {
    files: [],
    rejectedFiles: [],
    error: "",
  });
```

Reply:
Challenge accepted!